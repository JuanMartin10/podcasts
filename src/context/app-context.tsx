import { createContext, useContext, useState } from 'react';
import { useEntries } from '../hooks/useEntries';
import { type AppEntry } from '../models/types';

export interface AppContextTypes {
  loading: boolean;
  entries: AppEntry[] | undefined;
  currentPodcast: AppEntry | undefined;
  getEntries: () => Promise<void>;
  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
  setEntries: React.Dispatch<React.SetStateAction<AppEntry[] | undefined>>;
  setCurrentPodcast: React.Dispatch<React.SetStateAction<AppEntry | undefined>>;
}

interface ContextProps {
  children: React.ReactNode;
}

const AppContext = createContext<AppContextTypes | undefined>(undefined);

export const AppProvider: React.FC<ContextProps> = ({ children }) => {
  const [loading, setLoading] = useState(false);
  const [currentPodcast, setCurrentPodcast] = useState<AppEntry>();

  const { entries, getEntries, setEntries } = useEntries({
    loading,
    setLoading,
  });

  return (
    <AppContext.Provider
      value={{
        loading,
        entries,
        currentPodcast,
        getEntries,
        setLoading,
        setEntries,
        setCurrentPodcast,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = (): AppContextTypes => {
  const context = useContext(AppContext);
  if (context == null) {
    throw new Error('useAppContext must be used within an AppWrapper');
  }
  return context;
};
