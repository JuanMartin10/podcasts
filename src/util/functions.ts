import { Feed } from '../models/api-types';

export const formatTime = (ms: number) => {
  let totalSeconds = Math.floor(ms / 1000);
  const hours = Math.floor(totalSeconds / 3600);
  totalSeconds %= 3600;
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = totalSeconds % 60;
  const secondsString = seconds.toString().padStart(2, '0');

  if (hours > 0) {
    return `${hours}:${minutes.toString().padStart(2, '0')}:${secondsString}`;
  } else {
    return `${minutes}:${secondsString}`;
  }
};

export const formatDate = (dateString: string) => {
  const date = new Date(dateString);

  const year = date.getFullYear();
  const month = (date.getMonth() + 1).toString().padStart(2, '0');
  const day = date.getDate().toString().padStart(2, '0');

  return `${day}/${month}/${year}`;
};

export const formatApiEntries = (apiFeed: Feed) =>
  apiFeed.entry.map(entry => ({
    id: entry.id.attributes['im:id'],
    title: entry['im:name'].label,
    artist: entry['im:artist'].label,
    image: entry['im:image'][0].label,
    summary: entry.summary.label,
  }));
