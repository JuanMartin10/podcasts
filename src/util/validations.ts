export const isEmpty = (val: string | null | undefined) => {
  return val === '' || val === null || val === undefined;
};

export const isFalse = (val: boolean) => !val;
