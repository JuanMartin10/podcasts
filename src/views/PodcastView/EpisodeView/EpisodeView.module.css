.play {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  padding: 24px;
  gap: 24px;
  box-shadow: 0.12em 0.25em 0.5em 0.25em rgba(0, 0, 0, 0.1);
}

h3 {
  margin: 0;
  text-align: left;
}
