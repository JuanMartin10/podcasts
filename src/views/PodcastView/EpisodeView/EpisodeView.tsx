import { useLocation } from 'react-router-dom';
import { useAppContext } from 'context/app-context';
import DetailsLayout from 'layout/DetailsLayout/DetailsLayout';
import styles from './EpisodeView.module.css';

const EpisodeView = () => {
  const { currentPodcast } = useAppContext();
  const { state: episode } = useLocation();

  return (
    <DetailsLayout currentPodcast={currentPodcast}>
      <div className={styles.play}>
        <h3>{episode?.trackName}</h3>
        {Boolean(episode?.description) && (
          <p
            dangerouslySetInnerHTML={{
              __html: episode?.description as string,
            }}
          ></p>
        )}
        <audio controls>
          <source src={episode.episodeUrl} type='audio/mpeg' />
        </audio>
      </div>
    </DetailsLayout>
  );
};

export default EpisodeView;
