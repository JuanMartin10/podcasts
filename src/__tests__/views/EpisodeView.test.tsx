import { afterEach, beforeEach, describe, expect, it } from 'vitest';
import { render, screen, cleanup, waitFor } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { AppContext } from 'context/app-context';
import currentPodcastJson from 'mocks/podcast.json';
import episodesJson from 'mocks/episodes.json';
import { formatApiEntries } from 'util/functions';
import EpisodeView from 'views/PodcastView/EpisodeView/EpisodeView';

const renderWithProvider = (ui: React.ReactElement) =>
  render(
    <AppContext.Provider
      value={{
        entries: formatApiEntries(episodesJson.feed as any),
        loading: false,
        currentPodcast: currentPodcastJson.podcast,
        setLoading: () => {},
        setCurrentPodcast: () => {},
        setEntries: () => {},
      }}
    >
      {ui}
    </AppContext.Provider>
  );
const renderLoading = (ui: React.ReactElement) =>
  render(
    <AppContext.Provider
      value={{
        entries: formatApiEntries(episodesJson.feed as any),
        loading: true,
        currentPodcast: undefined,
        setLoading: () => {},
        setCurrentPodcast: () => {},
        setEntries: () => {},
      }}
    >
      {ui}
    </AppContext.Provider>
  );
const fakeEpisode = {
  wrapperType: 'track',
  releaseDate: '2023-03-31T02:54:00Z',
  trackTimeMillis: 3635,
  trackId: 1574029840,
  trackName: 'Friday Night Karaoke',
};
const wrapper = ({ id, episodeId, children }: any) => (
  <MemoryRouter
    initialEntries={[
      `/podcast/${id}/episode/${episodeId}`,
      { state: fakeEpisode },
    ]}
  >
    <Routes>
      <Route
        path='/podcast/:id/episode/:episodeId'
        Component={EpisodeView}
      ></Route>
    </Routes>
  </MemoryRouter>
);

describe('<EpisodeView />', () => {
  const wrapperId = () =>
    wrapper({
      podcastId: '1574029840',
      episodeId: '1000608200247',
    });

  beforeEach(() => {
    renderWithProvider(wrapperId());
  });
  afterEach(cleanup);

  it('Should render', () => {
    renderWithProvider(wrapperId());
  });

  it("Should render 'Loading...'", () => {
    renderLoading(wrapperId());

    waitFor(() => {
      expect(screen.getByText(/load/i)).toBeDefined();
    });
  });

  it('Should match the snapshot', () => {
    const { container } = renderWithProvider(wrapperId());
    waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('Should render description episode', () => {
    waitFor(() => {
      const text = screen.getAllByText(/Episode 48 /i);

      expect(text).toBeDefined();
    });
  });

  it('Should render episode trackname as h3', () => {
    waitFor(() => {
      const heading = screen.getByRole('heading');

      expect(heading).toHaveTextContent('friday night karaoke');
    });
  });

  it('Should render one podcast image', async () => {
    waitFor(() => {
      expect(screen.getAllByRole('img')).toHaveLength(1);
    });
  });

  it('Should render description', () => {
    waitFor(() => {
      const description = screen.getByText(
        /No ads, no gimmicks - just Karaoke! Friday Night/i
      );
      expect(description).toBeDefined();
    });
  });

  it('Should render an audio player', () => {
    waitFor(() => {
      expect(screen.getByRole('audio')).toHaveLength(1);
    });
  });
});
