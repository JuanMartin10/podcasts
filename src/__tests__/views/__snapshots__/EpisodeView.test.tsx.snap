// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EpisodeView /> > Should match the snapshot 1`] = `<div />`;
